File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> lRes;
        string lTemp;
        int lOpn = 0, lClose = 0;
        backTracking(lRes, lTemp, lOpn, lClose, n);
        return lRes;
    }

    void backTracking(
        vector<string>& aInRes,
        string aInTemp,
        int aInOpn,
        int aInClose,
        int aInN
    )
    {
        if (aInTemp.size() == 2 * aInN)
        {
            if (find(aInRes.begin(), aInRes.end(), aInTemp) == aInRes.end())
            {
                aInRes.push_back(aInTemp);
            }
        }
        if (aInOpn < aInN)
        {
            backTracking(aInRes, aInTemp + '(', aInOpn + 1, aInClose, aInN);
        }
        if (aInClose < aInOpn)
        {
            backTracking(aInRes, aInTemp + ')', aInOpn, aInClose + 1, aInN);
        }
    }
};

File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        
        def backTracking(res, temp, opn, close):
            if len(temp) == 2 * n:
                if temp not in res:
                    res.append(temp)
            if opn < n:
                backTracking(res, temp + "(", opn + 1, close)
            if close < opn:
                backTracking(res, temp + ")", opn, close + 1)

            return res
        return backTracking([], "", 0, 0)